util/pkg/vfs/vaultcontext.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vfs
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sts"
+	vault "github.com/hashicorp/vault/api"
+)
+
+func newVaultClient(addr string) (*vault.Client, error) {
+
+	config := &vault.Config{
+		Address:    addr,
+		HttpClient: &http.Client{},
+	}
+	client, err := vault.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return client, err
+}
+
+func authenticateVaultClient(client *vault.Client) error {
+	token := os.Getenv("VAULT_TOKEN")
+	if token != "" {
+		client.SetToken(token)
+	} else {
+		return awsAuth(client)
+
+	}
+	return nil
+}
+
+func awsAuth(client *vault.Client) error {
```

Comment:
I wonder if we could use fi.Cloud here? or something that would allow the aws client to be mocked more easily.

Reply:
fi.Cloud is new to me. I'll have a look.

Reply:
Ah doh. The problem with fi.Cloud is that it causes import cycles. I cannot make use of anything from fi or from apis. So it gets a bit tricky to test for cloud provider.

Reply:
An in-package unit test could stuff a mock inside a field added to the `VFSContext` struct.